src/layers/icmpv6.rs

```diff
@@ -176,28 +222,87 @@ impl Layer for ICMPv6 {
                 let retrans_timer: u32 = u32::from_be_bytes(bytes[12..16].try_into().unwrap());
-                decoded = 16;
+
+                let option_type: u8 = u8::from_be(bytes[16]);
+                let length = u8::from_be(bytes[17]);
+                decoded = 17 + ((length << 3) as usize) - 2;
+                let src_link_layer = Icmpv6LinkLayerAddress {
+                    option_type,
+                    length,
+                    link_layer_address: bytes[18..decoded + 1].try_into().unwrap(),
+                };
+                decoded += 1;
+                let mtu = Icmpv6OptionsMTU {
+                    option_type: u8::from_be(bytes[decoded]),
+                    length: u8::from_be(bytes[decoded + 1]),
+
+                    mtu: u32::from_be_bytes(
+                        bytes[(decoded + 4)..(decoded + 8)].try_into().unwrap(),
+                    ),
+                };
+                decoded += 8;
+                let data_offset = decoded;
+                let prefix_information = Icmpv6PrefixInfo {
+                    option_type: u8::from_be(bytes[data_offset]),
+                    length: u8::from_be(bytes[data_offset + 1]),
+                    prefix_len: u8::from_be(bytes[data_offset + 2]),
+                    flag: Icmpv6OptionsFlag {
+                        on_link_flag: ((bytes[data_offset + 3] >> 7) & 0x01) == 0x01,
+                        autonomous_address_congif_flag: ((bytes[data_offset + 3] >> 7) & 0x01)
+                            == 0x01,
+                    },
+                    valid_lifetime: u32::from_be_bytes(
+                        bytes[(data_offset + 4)..(data_offset + 8)]
+                            .try_into()
+                            .unwrap(),
+                    ),
+                    preferred_lifetime: u32::from_be_bytes(
+                        bytes[(data_offset + 8)..(data_offset + 12)]
+                            .try_into()
+                            .unwrap(),
+                    ),
+                    prefix: bytes[(data_offset + 16)..(data_offset + 32)]
+                        .try_into()
+                        .unwrap(),
+                };
+                decoded += 32;
+
                 Icmpv6Type::RouterAdvertisement(Icmpv6RouterAdvertisement {
                     cur_hop_limit,
                     flags,
                     router_lifetime,
                     reachable_time,
                     retrans_timer,
+                    src_link_layer,
+                    mtu,
+                    prefix_information,
                 })
             }
 
             ICMPV6_NEIGHBOR_SOLICITATION => {
                 let target_address = bytes[8..24].try_into().unwrap();
-                decoded = 24;
-                Icmpv6Type::NeighborSolicitation(Icmpv6NeighborSolicitation { target_address })
+                let option_type: u8 = u8::from_be(bytes[24]);
```

Comment:
This and remaining code can be properly reviewed after the above structure changes are done.